### PR TITLE
bartender: init at 4.2.21

### PR DESCRIPTION
**Pull request by stepbrobd**
###### Description of changes

added new darwin only package, bartender <https://www.macbartender.com/>

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by StepBroBD** [comment]
On aarch64-darwin:

```shell
$ nix run nixpkgs#nixpkgs-review -- pr 240618
1 package added:
bartender (init at 4.2.21)
1 package built:
bartender
```

File: pkgs/os-specific/darwin/bartender/default.nix [hunk 1]
```diff
@@ -0,0 +1,48 @@
+{ lib
+, stdenvNoCC
+, fetchurl
+, undmg
+}:
+
+stdenvNoCC.mkDerivation rec {
```
Comment (drupol) [inline pkgs/os-specific/darwin/bartender/default.nix:7]: I think we should use the `finalAttrs` pattern here.

Let me share a couple of useful links regarding this:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

File: pkgs/os-specific/darwin/bartender/default.nix [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  pname = "bartender";
+  version = "4.2.21";
+
+  src = fetchurl {
+    name = "Bartender 4.dmg";
+    url = "https://www.macbartender.com/B2/updates/${builtins.replaceStrings [ "." ] [ "-" ] version}/Bartender%204.dmg";
+    sha256 = "sha256-KL4Wy8adGiYmxaDkhGJjwobU5szpW2j7ObgHyp02Dow=";
```
Comment (drupol) [inline pkgs/os-specific/darwin/bartender/default.nix:14]:
```suggestion
    hash = "sha256-KL4Wy8adGiYmxaDkhGJjwobU5szpW2j7ObgHyp02Dow=";
```

**Review by drupol — COMMENTED** [review]
Hello,

Thanks for your PR !

I added 2 non-blocking comments, it would be nice if you could implement them.
If not, I'll merge this PR during my next visit.

Thanks in advance!

**Review by natsukium — APPROVED** [review]
It works fine on aarch64-darwin.
LGTM.

**Review by drupol — APPROVED** [review]
LGTM, let's wait the end of the eval before merging.

Ping me when it's done so I don't have to watch this page.

Thanks.

**Comment by StepBroBD** [comment]
Eval and tests are done, feel free to merge at your convenience.

Related:
- https://github.com/NixOS/nixpkgs/pull/240625
- https://github.com/NixOS/nixpkgs/pull/240621
- https://github.com/NixOS/nixpkgs/pull/240618

**Pull request merged**